### PR TITLE
Fix to show build pipeline parse error

### LINKED ISSUE #811: Yaml parse issues are only reported back to hook and not shown in the UI
### Component

server

### Describe the bug

When a error exists in the commited yaml file this information is set back to the source control webhook... The build then never shows up and one has to check the source control webhook log to debug the issue...

One could argue that this is ok, but we could also just create a build that shows as failed and displays the error ? this to me seems like the expected behavior actually....

What do you think ? 

### System Info

```shell
Currently running master on my mac, agent also running right next to it

SCM: Gitea @ git.lbsfilm.at
```

### Additional context

_No response_

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A join our [Discord Chat Server](https://discord.gg/fcMQqSMXJy).

### LINKED ISSUE #1045: Improve Visibility of Errors on Invalid CI Config Yaml
### Clear and concise description of the problem

I am using Woodpecker with Gitea using webhooks to trigger CI runs on every push/commit.

Currently, if a commit contains invalid CI yaml, it silently fails to run the CI job, unless you look at the webhook configuration (which may not be available to lower-privileged users) and is not obvious unless you know what you're looking for.

The following screenshot was taken from `https://gitea.example.com/[project]/[repo]/settings/hooks/[hook-id]`:

image

However, if you're not sure what you're looking for, it looks like Ci just silently failed.

### Suggested solution

I would propose that a far more self-evident solution would be creating a job entry in the CI jobs list, but marking it as failed, and also providing failed feedback to the git repo to turn the build red, etc. This fits far more closely in the mental model of how a developer would expect a CI system to interact: fail the build if it couldn't complete successfully due to bad files in the repo.

This is similar to how gitlab CI approaches the problem. Below is an example of one of my projects - the bottom entry is a normal successful run, and then I intentionally committed invalid yaml, which generated the top entry in CI run entries.

image

Similarly, it's treated like any other failed job, and marks the commit as having had CI not pass (instead of gitea's solution of not creating any CI runs in the interface).

image

The run details has more specific information as well.

image

### Alternative

_No response_

### Additional context

_No response_

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by lafriks**
Also handle if no pipeline config is found in repository

Replaces #812

Fixes #811

Fixes #1045

Screenshots:
attels

attels

### CODE REVIEW

**Comment by lafriks** [comment]
Also fixes to handle restart of such jobs

File: server/pipeline/restart.go [hunk 1]
```diff
@@ -85,6 +87,10 @@ func Restart(ctx context.Context, store store.Store, lastBuild *model.Build, use
 
 	newBuild, buildItems, err := createBuildItems(ctx, store, newBuild, user, repo, pipelineFiles, envs)
 	if err != nil {
+		var parseErr *yaml.PipelineParseError
+		if errors.As(err, &parseErr) {
```
Comment (6543) [inline server/pipeline/restart.go:91]: why not .Is ?

Comment (lafriks) [reply server/pipeline/restart.go:91]: I don't think you can check typed errors with fields with `Is` 🤔

Comment (6543) [reply server/pipeline/restart.go:91]: did not looked up this ... I'll check

**Comment by lafriks** [comment]
Added to display error in UI if no pipeline config file is found

**Comment by 6543** [comment]
please merge master into

File: server/pipeline/restart.go (hunk 1 above)
Comment (6543) [reply server/pipeline/restart.go:99]: "... sets target to that error value and returns true" ...

we should not use As at all ...

Comment (6543) [reply server/pipeline/restart.go:99]: https://go.dev/play/p/zcJL_wTcC0q

it will always be true ...

File: pipeline/frontend/yaml/error.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+package yaml
+
+// PipelineParseError is an error that occurs when the pipeline parsing fails.
+type PipelineParseError struct {
+	Err error
+}
+
+func (e PipelineParseError) Error() string {
+	return e.Err.Error()
+}
```
Comment (6543) [inline pipeline/frontend/yaml/error.go:24]:
```suggestion
}

// IsPipelineParseError checks if an error is a PipelineParseError
func IsPipelineParseError(err error) bool {
	_, ok1 := err.(PipelineParseError)
	_, ok2 := err.(*PipelineParseError)
	return ok1 || ok2
}
```

File: server/pipeline/restart.go [hunk 3]
```diff
@@ -85,6 +95,10 @@ func Restart(ctx context.Context, store store.Store, lastBuild *model.Build, use
...
```
Comment (6543) [inline server/pipeline/restart.go:99]:
```suggestion
		if yaml.IsPipelineParseError(err) {
```

File: pipeline/frontend/yaml/error.go (hunk 2 above)
Comment (6543) [reply pipeline/frontend/yaml/error.go:24]:
```go
// IsPipelineParseError checks if an error is a PipelineParseError
func IsPipelineParseError(err error) bool {
	_, ok := err.(*PipelineParseError)
	return ok
}
```
would do it too ... if we are sure we only use &PipelineParseError{} version ...

Comment (lafriks) [reply pipeline/frontend/yaml/error.go:24]: Why not just use errors.As? As that would work also for wrapped errors

Comment (6543) [reply pipeline/frontend/yaml/error.go:24]: https://pkg.go.dev/errors#As

Comment (6543) [reply pipeline/frontend/yaml/error.go:24]: https://go.dev/play/p/pT0g50DWTku vs https://go.dev/play/p/2X1yEErc5XO I would say ...

in the end if this is the only blocking thing I'm fine

Comment (lafriks) [reply pipeline/frontend/yaml/error.go:24]: changed to `errors.Is`: https://go.dev/play/p/hpv3yhmlvzk

**Review by 6543 — APPROVED** [review]

**Pull request merged**